File: clinical-schema/python/src/dbbest_clinical_schema/models/visit.py
"""Visit model — mirrors VISIT_DIMENSION."""
from __future__ import annotations

from typing import Optional, Union

from pydantic import Field, field_validator

from ._base import ClinicalBase, iso_date
from ..constants import ACTIVE_STATUS_ACTIVE, InOut, DEFAULT_SOURCESYSTEM_CD
import json


class Visit(ClinicalBase):
    ENCOUNTER_NUM: int
    PATIENT_NUM: int
    START_DATE: str
    END_DATE: Optional[str] = None
    INOUT_CD: str = InOut.OUTPATIENT.value
    LOCATION_CD: Optional[str] = None
    ACTIVE_STATUS_CD: str = ACTIVE_STATUS_ACTIVE
    VISIT_BLOB: Optional[str] = None
    UPDATE_DATE: Optional[str] = None
    DOWNLOAD_DATE: Optional[str] = None
    IMPORT_DATE: Optional[str] = None
    SOURCESYSTEM_CD: str = DEFAULT_SOURCESYSTEM_CD
    UPLOAD_ID: int = 1
    CREATED_AT: str = Field(default_factory=iso_date)

    @field_validator("VISIT_BLOB", mode="before")
    @classmethod
    def _stringify_blob(cls, v: Union[str, dict, list, None]) -> Optional[str]:
        if v is None:
            return None
        if isinstance(v, (dict, list)):
            return json.dumps(v)
        return v

File: clinical-schema/js/src/builders/visit.js
import { ACTIVE_STATUS, INOUT, DEFAULT_SOURCESYSTEM_CD } from '../constants.js'
import { isoDate, stringifyBlob } from './util.js'

/**
 * Build a VISIT_DIMENSION-shaped record.
 * Required: ENCOUNTER_NUM, PATIENT_NUM, START_DATE.
 */
export function buildVisit({
  ENCOUNTER_NUM,
  PATIENT_NUM,
  START_DATE,
  END_DATE = null,
  INOUT_CD = INOUT.OUTPATIENT,
  LOCATION_CD = null,
  ACTIVE_STATUS_CD = ACTIVE_STATUS.ACTIVE,
  VISIT_BLOB = null,
  SOURCESYSTEM_CD = DEFAULT_SOURCESYSTEM_CD,
  UPLOAD_ID = 1,
} = {}) {
  if (ENCOUNTER_NUM == null) throw new Error('buildVisit: ENCOUNTER_NUM is required')
  if (PATIENT_NUM == null) throw new Error('buildVisit: PATIENT_NUM is required')
  if (!START_DATE) throw new Error('buildVisit: START_DATE is required')

  return {
    ENCOUNTER_NUM,
    PATIENT_NUM,
    ACTIVE_STATUS_CD,
    START_DATE,
    END_DATE,
    INOUT_CD,
    LOCATION_CD,
    VISIT_BLOB: stringifyBlob(VISIT_BLOB),
    UPDATE_DATE: null,
    DOWNLOAD_DATE: null,
    IMPORT_DATE: null,
    SOURCESYSTEM_CD,
    UPLOAD_ID,
    CREATED_AT: isoDate(),
  }
}
